classes/DataClass/DataFrame.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MySociety\TheyWorkForYou\DataClass;
+
+class Cell {
+    public string $key;
+    public $value;
+
+    public function getValue() {
+        return $this->value;
+    }
+}
+
+class Row {
+    /** @var Cell[] $cells */
+    public array $cells = [];
+
+    public function __construct(array $data) {
+        foreach ($data as $key => $value) {
+            $cell = new Cell();
+            $cell->key = (string) $key;
+            $cell->value = $value;
+            $this->cells[$key] = $cell;
+        }
+    }
+
+    public function get(string $column) {
+        $cell = $this->cells[$column] ?? null;
+        return $cell ? $cell->getValue() : null;
+    }
+
+    public function toArray(): array {
+        $array = [];
+        foreach ($this->cells as $cell) {
+            $array[$cell->key] = $cell->value;
+        }
+        return $array;
+    }
+}
+
+class DataFrame {
+    /**
+     * @var Row[] $rows An array of Row objects
+     * @var string[] $columns An array of strings representing column names
+     */
+    public array $rows = [];
+    public array $columns = [];
+
+    public function __construct(array $data) {
+        /**
+         * Accepts an array of associative arrays, where each associative array represents a row in the DataFrame.
+         * From this extract the column names and create a Row object for each row.
+         */
+        foreach ($data as $rowData) {
+            // update the columns with any new keys in each row
+            $this->columns = array_unique(array_merge($this->columns, array_keys($rowData)));
+            // make sure all columns are strings rather than ints
+            $this->rows[] = new Row($rowData);
+        }
+    }
+
+    public function toArray(): array {
+        return array_map(fn($row) => $row->toArray(), $this->rows);
+    }
+
+    public function toHTML(?string $url_column = null): string {
```

Comment:
I think I would prefer not to have things generated like this because it makes it harder to style the data and in this case the election_register template is quite hard to understand where the content is coming from and how to style it. Probably not for now but I think we should probably change this in the future to move this into the templates.

Reply:
Yeah this was just quickly moving it from the prototype (which was in python, so just used the pandas grids.)